### PR TITLE
Skip DB writes for unchanged Dags after serialization

### PR DESCRIPTION
**Pull request by ephraimbuddy**
Reduces metadata DB churn by skipping serialized DAG writes when a Dag has not changed, while still updating Dag model metadata and permissions to keep scheduling state current.

---

##### Was generative AI tooling used to co-author this PR?

- [x] Yes (please specify the tool below)

Generated-by: gpt-5.2-codex on cursor IDE
Generated-by:  Claude 4.5 Opus on cursor-agent cli

### CODE REVIEW

**Review by shahar1 — COMMENTED** [review]
Could you please try adding tests for checking behvaior of exceptions and handling the edge cases?

**Comment by shahar1** [comment]
mypy is unhappy :(

File: airflow-core/src/airflow/dag_processing/collection.py [hunk 1]
```diff
@@ -245,6 +248,190 @@ def _serialize_dag_capturing_errors(
...
+    if not dags:
+        return
+    log.debug("Updating DagModel for unchanged Dags", count=len(dags))
+    dag_op = DagModelOperation(
+        bundle_name=bundle_name,
+        bundle_version=bundle_version,
+        dags={dag.dag_id: dag for dag in dags},
+    )
+    orm_dags = dag_op.add_dags(session=session)
+    dag_op.update_dags(orm_dags, parse_duration, session=session)
```
Comment (uranusjr) [inline airflow-core/src/airflow/dag_processing/collection.py]: If the dag is unchanged, we can probably skill these altogether since the DagModel fields shouldn’t change eeither?

Comment (ephraimbuddy) [reply airflow-core/src/airflow/dag_processing/collection.py]: It's to avoid marking it as stale. Also some scheduling information depends on this such as calculating next dagrun date fields

Comment (ashb) [reply airflow-core/src/airflow/dag_processing/collection.py]: If that's all it is, it might be worth an explicit update just to the last parsed column? (Not sure)

Comment (ephraimbuddy) [reply airflow-core/src/airflow/dag_processing/collection.py]: I have implemented updates for specific fields.

File: airflow-core/src/airflow/dag_processing/collection.py [hunk 2]
```diff
@@ -404,6 +598,16 @@ def update_dag_parsing_results_in_db(
     # Retry 'DAG.bulk_write_to_db' & 'SerializedDagModel.bulk_sync_to_db' in case
     # of any Operational Errors
     # In case of failures, provide_session handles rollback
+    dags_list = list(dags)
```
Comment (uranusjr) [inline airflow-core/src/airflow/dag_processing/collection.py:601]: I checked the code, `update_dag_parsing_results_in_db` is actually always called with a list, so let’s just change the function annotation to `Sequence[LazyDeserializedDAG]` instead and remove `dags_list`.

**Review by uranusjr — COMMENTED** [review]
One comment otherwise I think this should work.
